src/foam/dao/CSVSink.js

```diff
@@ -76,34 +127,95 @@ foam.CLASS({
             Null: function(value) {}
           }, function(value) {
             this.output_(value.toString());
-          })
+        }),
+      javaCode: `
+        switch (value) {
+          case value == null: 
+            break;
+          case value instanceof String:
+            setCsv(getCsv() +  value.replace("\\"", "\\"\\""));
+            break;
+          case value instanceof Number:
+            setCsv(getCsv() + value.toString());
+            break;
+          case value instanceof Boolean: 
+            setCsv(getCsv() + value.toString());
+            break;
+          case value instanceof Date:
+            output_(value.toDateString());
+            break;
+          case value instanceof Array:
+            output_(foam.json.Pretty.stringify(value));
+            break;
+          case value instanceof FObject: 
+            output_(foam.json.Pretty.stringify(value));
+            break;
+          default:
+            output_(value.toString());
+        }
+      `
     },
-
-    function newLine_() {
-      this.csv += '\n';
-      this.isNewLine = true;
+    {
+      name: 'newLine_',
+      code: function() {
+        this.csv += '\n';
+        this.isNewLine = true;
+      },
+      javaCode: `
+        setCsv(getCsv() + "\\n");
+        setIsNewLine(true);
+      `
     },
+    {
+      name: 'put',
+      args: [
+        { class: 'FObject', name: 'obj' }
+      ],
+      code: function(obj) {
+        if ( ! this.of ) this.of = obj.cls_;
 
-    function put(obj) {
-      if ( ! this.of ) this.of = obj.cls_;
+        if ( ! this.isHeadersOutput ) {
+          this.props.forEach((element) => {
+            element.toCSVLabel(this, element);
+          });
+          this.newLine_();
+          this.isHeadersOutput = true;
+        }
 
-      if ( ! this.isHeadersOutput ) {
         this.props.forEach((element) => {
-          element.toCSVLabel(this, element);
+          element.toCSV(obj, this, element);
         });
         this.newLine_();
-        this.isHeadersOutput = true;
-      }
+      },
+      javaCode: `
+        if ( ! isPropertySet(getOf()) ) setOf(obj.getClassInfo());
+        FObjectArray props = getProps();
+        if ( ! getIsHeadersOutput() ) {
+          for (FObject element : props) {
+            element.javaToCSVLabel(this, element);
+          }
+          newLine_();
+          setIsHeadersOutput(true);
+        }
 
-      this.props.forEach((element) => {
-        element.toCSV(obj, this, element);
-      });
-      this.newLine_();
+        for (FObject element : props) {
+          element.javaToCSV(obj, this, element);
```

Comment:
Shouldn't need to pass in element, as the "this" of the javaToCSV method will be the elemen.

Reply:
should "this" not be the outputter(csvSink)?